src/Scope/GetScopeAtNode.php

```diff
@@ -0,0 +1,57 @@
+<?php
+declare(strict_types=1);
+
+namespace LanguageServer\Scope;
+
+use LanguageServer\DefinitionResolver;
+use Microsoft\PhpParser\ClassLike;
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\SourceFileNode;
+use Microsoft\PhpParser\Node\Statement\FunctionDeclaration;
+
+/**
+ * Returns the scope at the start of $node.
+ */
+function getScopeAtNode(DefinitionResolver $definitionResolver, Node $targetNode): Scope
+{
+    /** @var SourceFileNode The source file. */
+    $sourceFile = $targetNode->getRoot();
+    /** @var FunctionDeclaration|null The first function declaration met, excluding anonymous functions. */
+    $nearestFunctionDeclarationParent = $targetNode->getFirstAncestor(FunctionDeclaration::class);
+    /** @var ClassLike|null The first class met. */
+    $nearestClassLike = $targetNode->getFirstAncestor(ClassLike::class);
+
+    $traverser = new TreeTraverser($definitionResolver);
+    $resultScope = null;
+    $traverser->traverse(
+        $sourceFile,
```

Comment:
I don't really understand why you would want to traverse the whole source file from the root if you already have the node and every node is linked to its parent. This seems very inefficient, as you will visit a lot of nodes that you don't care about. A scope is intuitively definded as the the "area